src/ncp/ncp_base_mtd.cpp

```diff
@@ -4552,6 +4552,35 @@ template <> otError NcpBase::HandlePropertySet<SPINEL_PROP_PHY_PCAP_ENABLED>(voi
     return error;
 }
 
+template <> otError NcpBase::HandlePropertyGet<SPINEL_PROP_NET_LEAVE_GRACEFULLY>(void) { return OT_ERROR_NONE; }
+
+template <> otError NcpBase::HandlePropertySet<SPINEL_PROP_NET_LEAVE_GRACEFULLY>(void)
+{
+    otError error = OT_ERROR_NONE;
+    bool    leave;
+
+    SuccessOrExit(error = mDecoder.ReadBool(leave));
+
+    error = otThreadDetachGracefully(mInstance, ThreadDetachGracefullyHandler, this);
+
+exit:
+    return error;
+}
+
+void NcpBase::ThreadDetachGracefullyHandler(void *aContext)
+{
+    static_cast<NcpBase *>(aContext)->ThreadDetachGracefullyHandler();
+}
+
+void NcpBase::ThreadDetachGracefullyHandler(void)
+{
+    spinel_status_t status = SPINEL_STATUS_THREAD_DETACH_GRACEFULLY_DONE;
+
+    IgnoreError(otInstanceErasePersistentInfo(mInstance));
+    mChangedPropsSet.AddLastStatus(status);
+    mUpdateChangedPropsTask.Post();
```

Comment:
An alternative approach could be to emit `VALUE_IS` for `SPINEL_PROP_NET_LEAVE_GRACEFULLY` (perhaps with `false` as the value) to signal the completion of a "graceful detach".

This model is used by other commands with asynchronous end of operations, such as "scan" requests. The end of a scan is reported by sending an unsolicited `SPINEL_PROP_MAC_SCAN_STATE` update.

Reply:
This is better. Have updated to use this approach.

Currently I didn't add a value for `SPINEL_PROP_NET_LEAVE_GRACEFULLY` because it seems unnecessary. Since in the NcpSpinel we know that this is a notification (instead of the response for SetProperty), we know it means the completion of a graceful detach.